File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        LinkedList<Character> q = new LinkedList<>(); 
        boolean answer = true;

        for (int i = 0; i < s.length(); i++) {
            if (q.isEmpty()) {
                q.add(s.charAt(i));
            } else {
                if (s.charAt(i) == ')') {
                    if (q.peekLast() == '(') {
                        q.removeLast();
                    } else {
                        q.add(s.charAt(i));
                    }
                } else {
                    q.add(s.charAt(i));
                }
            }
        }
        
        if (!q.isEmpty()) {
            answer = false;
        }
        return answer;
    }
}

File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.py
from collections import deque

def solution(s):
    answer = True
    
    q = deque()
    for i in range(len(s)):
        if len(q) == 0:
            q.append(s[i])
        else:
            if s[i] == "(":
                q.append(s[i])
            elif q[-1] == "(" and s[i] == ")":
                q.pop()
    
    if q:
        answer = False

    return answer